3.7 - Add curl based Http\Client Adapter.

markstory opened the pull request:
The stream based client has a few limitations around `fopen_url` settings, and proxy issues (#10841 ) that are not currently solvable. By adding a cURL based adapter we can default to using this extension which solves these problems and allows more feature support in the future.

### TODO

* [x] Verify results on timeout expiration.

ADmad committed:
Add interface for http client adapters.

ADmad committed:
Add adapter instance type check.

ADmad committed:
Add stub class for Curl adapter.

Make Http/Client use Curl adapter by default if "curl" extension is loaded.

markstory committed:
Start implementing options and tests for Curl adapter.

markstory committed:
Start implementing SSL options.

markstory committed:
Finish basic implementation and add live network tests.

markstory committed:
Handle DNS failure.

ADmad commented on src/Http/Client/Adapter/Curl.php:
```diff
@@ -0,0 +1,143 @@
+            'timeout' => CURLOPT_TIMEOUT,
+            'ssl_verify_peer' => CURLOPT_SSL_VERIFYPEER,
+            'ssl_verify_host' => CURLOPT_SSL_VERIFYHOST,
+            'ssl_verify_status' => CURLOPT_SSL_VERIFYSTATUS,
```
Seems travis has an older Curl version and `CURLOPT_SSL_VERIFYSTATUS` is unavailable.
https://travis-ci.org/cakephp/cakephp/jobs/390291411#L770

PHP Manual says:
> Added in cURL 7.41.0. Available since PHP 7.0.7.

markstory replied:
Perhaps this should be moved into an additional option like @lorenzo suggested?

lorenzo commented on src/Http/Client/Adapter/Curl.php:
```diff
@@ -0,0 +1,143 @@
+        if (isset($options['proxy']['proxy'])) {
+            $out[CURLOPT_PROXY] = $options['proxy']['proxy'];
+        }
+
```
Setting extra curl opta is not uncommon. Should we have an extra key in he array for custom opts?

markstory replied:
Would `$options['curl']` work as a place to put the additional curl options?

lorenzo replied:
Yeah, that sounds good to me

lorenzo commented on src/Http/Client/Adapter/Curl.php:
```diff
@@ -0,0 +1,143 @@
+     *
+     * @param resource $handle Curl handle
+     * @param string $responseData string The response data from curl_exec
+     * @return \Cake\Http\Client\Response[]
```
How can you get more than one response?

markstory replied:
You can't. I'll fix this.

markstory committed:
Fix doc block

markstory committed:
Add support for custom curl options.

Add ability for users to send arbitrary curl options. The provided curl
options will overwrite any options set by the adapter.

markstory commented:
Arbitrary curl option support added.

BassemN commented:
Thanks allot @markstory & @ADmad for implementing curl. This is essential for all modern apps.

ADmad commented:
@BassemN Thanks for reporting the issue (on my plugin) :slightly_smiling_face:.

markstory committed:
Fix warning for verify peer options & add status codes.

Timeout errors in the client should raise 504 (gateway timeout) so that
client applications can respond accordingly.

markstory committed:
Skip tests when curl is missing & suggest it in composer.

markstory committed:
Fix mistake.

lorenzo commented on src/Http/Client/AdapterInterface.php:
```diff
@@ -0,0 +1,28 @@
+     *
+     * @param \Cake\Http\Client\Request $request The request object to send.
+     * @param array $options Array of options for the stream.
+     * @return \Cake\Http\Client\Response[] Array of populated Response objects
```
I think this returns a single Response, not many

lorenzo commented on src/Http/Client/Adapter/Curl.php:
```diff
@@ -0,0 +1,161 @@
+     *
+     * @param resource $handle Curl handle
+     * @param string $responseData string The response data from curl_exec
+     * @return \Cake\Http\Client\Response
```
The type here says it is a single response, but for some reason an array is returned at the end
